runtime/common/src/xcm_teleport.rs

```diff
@@ -55,35 +73,126 @@ where
             <R as pallet_evm::Config>::AddressMapping::into_account_id(handle.context().caller);
         let origin: O = frame_system::RawOrigin::Signed(account_id).into();
 
-        let destination = VersionedLocation::V5(L::get());
+        let TeleportCallParams {
+            destination,
+            beneficiary,
+            assets,
+            fee_asset_item,
+        } = Self::build_teleport_params(destination_account, amount)?;
 
-        let beneficiary = VersionedLocation::V5(Location::new(
+        let call: C = pallet_xcm::Call::<R>::teleport_assets {
+            dest: Box::new(VersionedLocation::V5(destination)),
+            beneficiary: Box::new(VersionedLocation::V5(beneficiary)),
+            assets: Box::new(VersionedAssets::V5(assets)),
+            fee_asset_item,
+        }
+        .into();
+
+        RuntimeHelper::<R>::try_dispatch::<C>(handle, origin, call, 0)?;
+
+        Ok(())
+    }
+
+    #[precompile::public("deliveryFee(bytes32,uint256)")]
+    fn delivery_fee(
+        handle: &mut impl PrecompileHandle,
+        destination_account: H256,
+        amount: U256,
+    ) -> EvmResult<U256> {
+        // No benchmarks availabe yet for precompiles, so charge some arbitrary gas as a spam
+        // prevention mechanism.
+        handle.record_cost(1000)?;
+
+        let TeleportCallParams {
+            destination,
+            beneficiary,
+            assets,
+            fee_asset_item,
+        } = Self::build_teleport_params(destination_account, amount)?;
+
+        let program = Self::teleport_assets_program(
+            destination.clone(),
+            beneficiary,
+            assets,
+            fee_asset_item,
+        )?;
+
+        let versioned_fees = pallet_xcm::Pallet::<R>::query_delivery_fees(
+            VersionedLocation::V5(destination),
+            VersionedXcm::V5(program),
+        )
+        .map_err(|_| revert("cannot query delivery fees"))?;
+
+        let fees: Assets = versioned_fees.try_into().unwrap();
+        match &fees.get(fee_asset_item as usize).unwrap().fun {
+            Fungibility::Fungible(amount) => Ok(U256::from(*amount)),
+            _ => unreachable!(),
```

Comment:
It is important that this code never panics.
Even if the panics are practically not reachable in the current configuration, I'd remove all these throwing instructions.